Don't allow to set up GL waveforms if GL is not available

daschuer opened the pull request:
This fixes a crasher when no GL is available, but GL functions are used anyway. #14561

ronso0 committed:
Waveform preferences: move get/set acceleration to factory methods

ronso0 committed:
Waveform preferences: simplify set/update methods

ronso0 committed:
Waveform preferences: remove empty destructor

ronso0 committed:
Waveform preferences: move get/set WaveformOptions to factory methods

ronso0 committed:
Waveform preferences: factory writes beatgrid alpha to config

daschuer committed:
Fall back to WaveformWidgetBackend::None if some other is configured but not available.

daschuer committed:
Rename pFactory

ronso0 commented:
Thanks for looking into this!

This PR overlaps with the changes of #15421 
Do you see a chance to backport that and build the fix on top?

ronso0 commented:
Is this maybe also fixing #15858 ??
At least the stacktrace shows some GL stuff

daschuer committed:
Merge remote-tracking branch 'upstream/2.6' into gh14561

acolombier reviewed (approved):
Code is looking good, thanks both for the good cleanup! Ran a few manual tests and everything still seem to function as expected. 

Let merge this in 2.6 and as this improve the GL situation with waveform,

daschuer committed:
Merge remote-tracking branch 'upstream/2.6' into gh14561

daschuer commented:
This is again ready for merge.

m0dB reviewed (approved):
LGTM

daschuer commented:
@m0dB merge? Maybe someone else wants to do it.

daschuer committed:
Always handle all WaveformWidgetBackend values

daschuer committed:
Verify in setAcceleration() that GL or GLES is available

daschuer committed:
Add overload setValue(const ConfigKey& key, QFlagsType& value)

daschuer committed:
Fix reading allshader::WaveformRendererSignalBase::Options as single Option

daschuer committed:
Introduce kBeatGridAlpha for de-duplication

daschuer commented:
Done

ronso0 commented:
clazy ant macOS now complain about congif enums :man_shrugging:

daschuer committed:
Reqires that the generic overload is not an enum type

ronso0 commented:
Now **all** runners stumle over that unfortunatly.

daschuer committed:
Merge remote-tracking branch 'upstream/2.6' into gh14561

daschuer commented:
conflicts resolved.

ronso0 commented:
LGTM but the configobject changes are a bit above my paygrade.
@mixxxdj/developers who can check the enum/overload commits?

daschuer commented:
Two weeks with now answer form @mixxxdj/developers have passed. So I went ahead and asked copilot if it is now satisfied. 
You find the responds below.  I will remove the semicolon and add the comment as suggested. 
 
Copilot responds: 

Short answer: Yes — the Copilot Review change requests appear to be implemented and the PR consistently centralizes GL availability and waveform option handling in WaveformWidgetFactory. I inspected the PR (files changed) and verified the key requested changes; below I summarize what I checked, what was fixed, and a few small remaining suggestions.

What I did

* Opened PR #15762 and fetched the changed files.
* Reviewed the diffs for the six modified files (notably: src/preferences/configobject.h, dlgprefwaveform.{cpp,h}, waveformwidgetfactory.{cpp,h}, waveformwidgettype.h).
* Verified that functionality and API changes are consistent across the UI (DlgPrefWaveform) and the factory (WaveformWidgetFactory).

PR summary / intent

* Title: "Don't allow to set up GL waveforms if GL is not available"

* Purpose: Avoid crashes and bad states when OpenGL is not available by centralizing backend selection, options and availability checks in WaveformWidgetFactory and removing direct, duplicated logic from the preferences dialog.

* Scope: 6 files changed, ~+221 / -186 lines. The changes:

  * Centralize waveform backend & options handling in WaveformWidgetFactory (new APIs).
  * Modify DlgPrefWaveform to call the factory API instead of manipulating config keys and GL details directly.
  * Add safer config templates (ConfigObject) overloads for enum vs QFlags types.
  * Persist beatGridAlpha via the factory/config consistently.
  * Prevent selecting GL-based backends when OpenGL is unavailable.

Core changes — what I verified

* Centralized logic in WaveformWidgetFactory:

  * New methods added and implemented:

    * setAcceleration(bool) — sets a backend in config based on availability.
    * getWaveformOptions(), getWaveformOptionsSupportedByType(...), setWaveformOption(...), resetWaveformOptions().

  * getBackendFromConfig() now normalizes chosen backend to None when GL/GLSL/AllShader aren’t available.

  * createXxxWaveformWidget() switch statements consistently fall back to software widgets when backend is not available.

  * setDisplayBeatGridAlpha() now writes the config key (kBeatGridAlpha).

* UI changes in DlgPrefWaveform:

  * Replaced several direct reads/writes of kHardwareAccelerationKey / waveform\_options with the new factory API (pFactory->...).
  * UI enabling/disable logic now consults the factory (pFactory->getBackendFromConfig(), pFactory->widgetTypeSupportsAcceleration(), etc.).
  * Methods' signatures in dlgprefwaveform.h updated to match simplified factory-driven functions.

* ConfigObject templates:
  * Added is\_qflags trait and template requires to disambiguate enum vs QFlags overloads. This prevents ambiguous overloads when serializing/deserializing QFlags.

* Minor naming/style: replaced local variable factory with pFactory for consistency.

Other changes and consistency checks

* Defaults for allshader::Options: code now constructs an Options from Option::None where needed (avoids type mismatch).
* Removed the declared (but empty) DlgPrefWaveform destructor from the header — ok if not needed.
* WaveformWidgetBackend enum: the patch removed a surrounding #ifdef around AllShader in waveformwidgettype.h so AllShader is always present. The implementation still checks capabilities with #ifdef MIXXX\_USE\_QOPENGL in places that must only be compiled when OpenGL support exists.
* Many dialog code paths now call pFactory->setWidgetTypeFromHandle(...) and then update UI via pFactory queries — this centralization looks correct.

Potential issues / suggestions (minor)

* waveformwidgettype.h: AllShader is now always present in the enum (the #ifdef around it was removed). This is probably fine, but it changes the enum across builds where MIXXX\_USE\_QOPENGL is not defined. If the intent was to keep the enum identical across build configurations, it’s fine; otherwise consider keeping the conditional compilation so enum values match expected build-time capabilities.
* getBackendFromConfig() function ends with "return backend;\n};" — an extra semicolon after a function definition is harmless but unusual. No functional problem, but you could remove the trailing semicolon for style/consistency.
* setAcceleration(bool) writes the backend into m\_config, but it does not directly trigger widget recreation/update. That’s fine everywhere I saw because callers typically follow with setWidgetTypeFromHandle(...) or update functions, but consider documenting that setAcceleration does not refresh existing widgets (or make it perform the refresh if that is desired).
* The new ConfigObject template constraints rely on C++20 requires/traits (std::is\_enum\_v and QFlags detection). Ensure the project build configuration supports C++20 for all targeted compilers and platforms — otherwise these will fail to compile.
* A quick grep (not executed here) for any remaining uses of hard-coded waveform option config keys in other files would be good to ensure there are no stray places still using the old keys directly.

Risk / readiness assessment

* Merge readiness: Low-to-medium risk.

  * The logic centralization reduces crash risk and duplicates, and most UI calls were updated accordingly.
  * The change touches both UI and factory code; run CI and a functional test of waveform prefs on platforms with and without OpenGL to confirm no regressions.
  * Recommend running the test matrix or manual check on a machine without GL (or with GL disabled) to validate the crash is fixed.

* No blocking code correctness issues found in the diffs reviewed; just stylistic/consistency notes above.

Concrete actionable items (what I recommend before merging)

* Run CI (build + unit tests) for both OpenGL-enabled and OpenGL-disabled configurations.
* Manual test: open preferences on systems with/without GL, toggle waveform types and acceleration, and ensure no crash / correct enabling/disabling of options.
* (Optional) Restore the #ifdef around AllShader in waveformwidgettype.h if you want enum consistency across build configs — or add a comment clarifying the intent.

daschuer commented:
Done

ywwg commented on src/waveform/waveformwidgetfactory.cpp:
```diff
@@ -695,6 +697,21 @@ bool WaveformWidgetFactory::setWidgetTypeFromHandle(int handleIndex, bool force)
 
+WaveformWidgetBackend WaveformWidgetFactory::setAcceleration(bool enabled) {
+    WaveformWidgetBackend backend = WaveformWidgetBackend::None;
+    if (enabled) {
+        backend =
+#ifdef MIXXX_USE_QOPENGL
+                WaveformWidgetBackend::AllShader
+#else
+                WaveformWidgetBackend::GL
+#endif
+                ;
+    }
+    m_config->setValue(kHardwareAccelerationKey, backend);
+    return backend;
+}
```
@daschuer what do you think about this comment?

daschuer replied:
I think this comment is outdated.

ywwg commented on src/waveform/waveformwidgetfactory.cpp:
```diff
@@ -1338,6 +1378,45 @@ void WaveformWidgetFactory::setDefaultBackend() {
     m_config->setValue(kHardwareAccelerationKey, preferredBackend());
 }
 
+allshader::WaveformRendererSignalBase::Options WaveformWidgetFactory::getWaveformOptions() {
+    auto options = m_config->getValue(
+            kWaveformOptionsKey,
+            allshader::WaveformRendererSignalBase::Option::None);
+    return options;
+}
```
here's another ai question

daschuer replied:
This comment is also outdated.

ywwg reviewed (commented):
the changes here are hard to read because it's c++, but I am confident that we can check the correctness of these changes with some manual QA under various circumstances.  Let's merge this and then if something comes up we can fix it.

there are a couple of ai review comments that are worthy of response, please let me know what you think

daschuer commented:
> there are a couple of ai review comments that are worthy of response, please let me know what you think

There comment are outdated and this PR is ready to merge.
